### PR TITLE
fix: ai agents4 pharma docker gpu fix

### PR DESCRIPTION
**Pull request by ansh-info**
# For authors
Screenshot 2025-04-12 at 20 13 25

## Description
The previous approach used `docker-compose` profiles and service-specific logic for launching CPU/NVIDIA/AMD variants of the Ollama container. However, this led to maintenance overhead, fragile synchronization, and issues with model readiness across platforms.

This PR introduces a streamlined, production-ready startup system that dynamically adapts to the host's hardware and ensures the Ollama model is fully available before starting the app.

**Key changes:**

- **Introduces a unified `startup.sh` script** that:    
    - Detects GPU type (NVIDIA, AMD, or CPU fallback).        
    - Selects the appropriate `ollama` image (`latest` or `rocm`).        
    - Launches `ollama` via `docker run` for GPU types with correct runtime and entrypoint settings.        
    - Falls back to `docker compose` for CPU.        
    - Waits for the Ollama API and pulls the `nomic-embed-text` model reliably.        
    - Starts `talk2aiagents4pharma` / `talk2knowledgegraphs` only after model availability.
        
- **Replaces multi-profile `docker-compose.yml`** with a unified configuration that works across all hardware without profile switching.
    
- **Simplifies `ollama` service definition**:   
    - Uses a single, minimal entrypoint for CPU (`command: ollama serve`).        
    - Network made external for compatibility with `docker run`.
        
- **Improves startup resilience**:
    
    - Handles network creation.        
    - Replaces reliance on healthchecks with direct API readiness check.
        

**Benefits:**

- One Click On-boarding for users just by using one command: 
```sh
chmod +x startup.sh
./startup.sh
```
- Cross-platform support: NVIDIA, AMD, and CPU work out of the box.  
- Eliminates healthcheck dependency inside Ollama containers.    
- Guarantees application doesn't start until model is fully loaded.
- Easier to maintain, test, and debug.
    

This update makes the setup fully production-ready with clean GPU detection, safe container orchestration, and synchronized model loading. 

## Major Update to the README.md as well for easier on-boarding for the users.   
-> Divided into three sections, thanks to the suggestion by @gurdeep330 
-> In the following order:
1. Docker (stable-release)
2. git (for developers and contributors)
3. pip (beta-release)

Please view the updated README.md using this [link](https://github.com/ansh-info/AIAgents4Pharma/tree/Fix/AIAgents4Pharma-Docker-Gpu-Fix)

## Fixes # (issue) Mention the issue number.
#129 

## Type of change
Please delete options that are not relevant.
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

## How Has This Been Tested?
Please describe the tests you conducted to verify your changes. These may involve creating new test scripts or updating existing ones.
- [ ] Added new test(s) in the ```tests``` folder
- [ ] Added new function(s) to an existing test(s) (e.g.: ```tests/testX.py```)
- [x] No new tests added (Please explain the rationale in this case)

## Checklist
- [x] My code follows the style guidelines mentioned in the Code/DevOps guides
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (e.g. MkDocs)
- [x] My changes generate no new warnings
- [x] I have added or updated tests (in the ```tests``` folder) that prove my fix is effective or that my feature works
- [x] New and existing tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

# For reviewers

## Checklist pre-approval
- [ ] Is there enough documentation?  
- [ ] If a new feature has been added, or a bug fixed, has a test been added to confirm good behavior?
- [ ] Does the test(s) successfully test edge/corner cases?
- [ ] Does the PR pass the tests? (if the repository has continuous integration)

## Checklist post-approval
- [ ] Does this PR merge ```develop``` into ```main```? If so, please make sure to add a prefix (feat/fix/chore) and/or a suffix BREAKING CHANGE (if it's a major release) to your commit message.
- [ ] Does this PR close an issue? If so, please make sure to descriptively close this issue when the PR is merged.

## Checklist post-merge
- [ ] When you approve of the PR, merge and close it (Read this [article](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/about-merge-methods-on-github) to know about different merge methods on GitHub)
- [ ] Did this PR merge ```develop``` into ```main``` and is it suppose to run an automated release workflow (if applicable)? If so, please make sure to check under the "Actions" tab to see if the workflow has been initiated, and return later to verify that it has completed successfully.

### CODE REVIEW

File: README.md
Comment (dmccloskey) [inline README.md:1]: @ansh-info  have you checked with @awmulyadi if the new docker images work with T2KG?

File: aiagents4pharma/talk2knowledgegraphs/startup.sh [hunk 1]
```diff
@@ -0,0 +1,88 @@
+#!/bin/bash
+set -e
+
+echo "[STARTUP] Detecting GPU type..."
```
Comment (dmccloskey) [inline aiagents4pharma/talk2knowledgegraphs/startup.sh:4]: I would recommend changing this to "Hardware configuration" to avoid any confusion. I would also recommend adding an option i.e., --cpu to force the use of the CPU even when a GPU is detected. This is quite useful when developing GPU code to check that it was implemented correctly by comparing to the CPU. Also, sometimes the user may just want to run on the CPU instead of the GPU even if they have one.

**Review by dmccloskey — CHANGES_REQUESTED** [review]
Nice working in finding a solution to detect the hardware config and deploying the appropriate docker version of Ollama 👍.

The solution works for local deployments (i.e., fine for this PR), but we will run into limitations when trying to deploy and scale in the cloud. I would recommend taking a look at Helm https://helm.sh/ and Kubernetes https://kubernetes.io/ for future reference. Also, an example for NVIDIA NIMS https://docs.nvidia.com/nim/large-language-models/latest/deploy-helm.html that demonstrates integration with OpenTelemetry and autoscaling.

Please let me know if you have any questions on my minor comments.

File: README.md [hunk 2]
```diff
@@ -23,110 +26,102 @@ Our toolkit currently consists of the following agents:
...
-4. Save the file.
+_To use **Talk2AIAgents4Pharma** / **Talk2KnowledgeGraphs**, you'll need a free NVIDIA API key. You can create an NVIDIA account and apply for free credits [here](https://build.nvidia.com/explore/discover)._  
+ _LangSmith support is optional. If you'd like to enable it, you can get your API key [here](https://docs.smith.langchain.com/administration/how_to_guides/organization_management/create_account_api_key)._
 
-To start the containers, run the following command:
+##### **3. Start the application**
+
+Run the startup script. It will:
+
+- Detect your GPU (NVIDIA, AMD, or CPU)
```
[the remark this answers is not in the document]
Comment (ansh-info) [reply README.md]: Thank you for the review!
Apple Metal acceleration (for Apple Silicon CPUs like M1/M2/M3) is only available outside containers on macOS.
Since Docker containers run inside a lightweight Linux VM and cannot access the Metal API, we treat Apple Silicon (arm64) as CPU mode inside Docker.
I have updated startup.sh to detect arm64 architecture and clearly log that Metal acceleration is unavailable when running inside containers.
Regarding Intel CPUs: SIMD optimizations (such as MLK, AVX, SSE) are automatically leveraged by the standard ollama/ollama:latest CPU image without requiring a special runtime or manual configuration.
No additional changes are needed for Intel-specific acceleration.

[the remark this answers is not in the document]
Comment (ansh-info) [reply README.md]: To get Apple Metal acceleration on Mac M3, the user must natively install Ollama on macOS.
Docker-based Ollama will always run in CPU mode on Macs.

File: aiagents4pharma/talk2knowledgegraphs/startup.sh (hunk 1 above)
Comment (ansh-info) [reply aiagents4pharma/talk2knowledgegraphs/startup.sh:4]: Thank you for the suggestion!

I have updated the startup.sh script to rename the detection log from "GPU Type" to "Hardware configuration" for better clarity, especially when CPU-only setups are involved.

Additionally, I have implemented a --cpu command-line flag. When passed, it forces the use of CPU mode even if a compatible GPU is detected.
This is how user can use the sh script :
```sh
chmod +x startup.sh
./startup.sh        # use --cpu flag if you want to run on CPU
```

This enhancement allows developers to easily compare CPU vs GPU execution during development and gives users the flexibility to manually choose CPU mode if preferred.

Both changes have been tested and validated. Thank you again for the helpful feedback!

File: README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (ansh-info) [reply README.md]: @dmccloskey also I have updated the line 77 in the README.md to look like this now, please let me know if its correct:

- Detect your hardware configuration (NVIDIA GPU, AMD GPU, or CPU). Apple Metal is unavailable inside Docker, and Intel SIMD optimizations are automatically handled without special configuration.

**Comment by ansh-info** [comment]
> Nice working in finding a solution to detect the hardware config and deploying the appropriate docker version of Ollama 👍.
> 
> The solution works for local deployments (i.e., fine for this PR), but we will run into limitations when trying to deploy and scale in the cloud. I would recommend taking a look at Helm https://helm.sh/ and Kubernetes https://kubernetes.io/ for future reference. Also, an example for NVIDIA NIMS https://docs.nvidia.com/nim/large-language-models/latest/deploy-helm.html that demonstrates integration with OpenTelemetry and autoscaling.
> 
> Please let me know if you have any questions on my minor comments.

Thank you for your feedback and the helpful suggestions!

I appreciate you pointing me toward Helm and Kubernetes — I agree that scaling in the cloud will require a more robust orchestration approach beyond local Docker deployments. I will definitely review the resources you shared, including the NVIDIA NIMs Helm deployment example with OpenTelemetry and autoscaling.

I will reach out if I have any questions. Thank you again for the guidance!

File: README.md
Comment (ansh-info) [reply README.md:1]: Thank you for the reminder. I have not directly checked with @awmulyadi  yet. However, I have thoroughly tested all containers, including `talk2knowledgegraphs`, on both my Apple Silicon M3 machine (CPU mode) and on my
 NVIDIA GPU server via SSH (both CPU and GPU modes). Everything appears to be working well across both environments.
Please let me know if you would like me to coordinate directly with @awmulyadi to confirm as well.

Below are some screenshots of me running `talk2knowledgegraphs` / `` on NVIDIA GPU:

> Logs -> It detects the gpu

image


> Model Loaded in GPU

image

> Output 

image

File: README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (dmccloskey) [reply README.md]: @ansh-info Good insight about Apple Metal availability through Docker. Good to know about SIMD support for Ollama. The updated README looks good 👍.

File: README.md
Comment (dmccloskey) [reply README.md:1]: @ansh-info Good work extensively testing it on your end. If you don't mind coordinating with @awmulyadi that would be great to make sure the changes run smoothly on his side and also if he has any feedback from his end.

**Review by dmccloskey — APPROVED** [review]
The updates to my comments look good 👍

File: README.md
Comment (ansh-info) [reply README.md:1]: Thank you for the feedback. I will surely coordinate with @awmulyadi and have already reached out to him in the Decision section. #172

**Comment by awmulyadi** [comment]
@ansh-info Hi, Ansh! Great work on the updates on dynamic scripts that support multiple platforms 👍. I have tested the scripts to run talk2knowledgegraphs on my Windows-based laptop successfully, as shown below.

Screenshot 2025-04-29 150158

However, I have several comments regarding my experience running the scripts (on Windows), as follows:
1. The `.env.example` under` talk2knowledgegraphs` has no `NVIDIA_API_KEY=...` as described in the readme file. Either add it as an additional line in the environment file or change the description in the readme file. 
Screenshot 2025-04-29 142314
2. The script to start the agent was not straightforward if you are using Windows (without WSL). In my case, I was able to run it by using the `Git Bash Here` terminal as suggested [here](https://stackoverflow.com/questions/26522789/how-to-run-sh-on-windows-command-prompt). We might need to include such additional information just in case the user came across the same issue as mine.
Screenshot 2025-04-29 142501
3. After installing Docker in windows, the user may need [an additional step](https://forums.docker.com/t/error-when-trying-to-run-windows-containers-docker-client-must-be-run-with-elevated-privileges/136619/6) to set it up before running the script to avoid the error messages as shown below. 
Screenshot 2025-04-29 144655

**Comment by ansh-info** [comment]
> 3\. [an additional step](https://forums.docker.com/t/error-when-trying-to-run-windows-containers-docker-client-must-be-run-with-elevated-privileges/136619/6)

Hi @awmulyadi,

Thank you for taking the time to run the application on your machine.😃 Regarding the points you mentioned:

1. I have added the `NVIDIA_API_KEY=...` in the `.env.example` file under **talk2knowledgegraphs**.
2. I have also updated the README to include instructions and a note for Windows users when running bash scripts.
3. I have added a note about the Docker setup requirements for Windows as well.


Could you please check if the updates in the README look fine,
Please let me know if everything looks good or if you would suggest any further improvements.

Thank you so much once again for your valuable feedback!

**Comment by awmulyadi** [comment]
> > 3. [an additional step](https://forums.docker.com/t/error-when-trying-to-run-windows-containers-docker-client-must-be-run-with-elevated-privileges/136619/6)
> 
> Hi @awmulyadi,
> 
> Thank you for taking the time to run the application on your machine.😃 Regarding the points you mentioned:
> 
> 1. I have added the `NVIDIA_API_KEY=...` in the `.env.example` file under **talk2knowledgegraphs**.
> 2. I have also updated the README to include instructions and a note for Windows users when running bash scripts.
> 3. I have added a note about the Docker setup requirements for Windows as well.
> 
> Could you please check if the updates in the README look fine, Please let me know if everything looks good or if you would suggest any further improvements.
> 
> Thank you so much once again for your valuable feedback!

The updated version of the README file includes useful notes for Windows users for running the script and looks good to me. Well done!👍

**Comment by ansh-info** [comment]
> > > 3. [an additional step](https://forums.docker.com/t/error-when-trying-to-run-windows-containers-docker-client-must-be-run-with-elevated-privileges/136619/6)
> > 
> > 
> > Hi @awmulyadi,
> > Thank you for taking the time to run the application on your machine.😃 Regarding the points you mentioned:
> > 
> > 1. I have added the `NVIDIA_API_KEY=...` in the `.env.example` file under **talk2knowledgegraphs**.
> > 2. I have also updated the README to include instructions and a note for Windows users when running bash scripts.
> > 3. I have added a note about the Docker setup requirements for Windows as well.
> > 
> > Could you please check if the updates in the README look fine, Please let me know if everything looks good or if you would suggest any further improvements.
> > Thank you so much once again for your valuable feedback!
> 
> The updated version of the README file includes useful notes for Windows users for running the script and looks good to me. Well done!👍

Thank you @awmulyadi, 😃

Hi @dmccloskey,
Just to update — @awmulyadi has tested the changes on his side and approved them.
When you have a chance, could you kindly review the PR and let me know if it’s good to proceed with the merge? 
Thank you!😃
